Refactor Aqara remote switch H1 quirk to v2 format to support all variants

jeverley opened the pull request:
## Proposed change
This PR refactors the existing Aqara H1 quirk into the new v2 structure (originally undertaken as I'd encountered a new device variant for the double switch):

```
"signature": {
  "node_descriptor": {
    "logical_type": 2,
    "complex_descriptor_available": 0,
    "user_descriptor_available": 0,
    "reserved": 0,
    "aps_flags": 0,
    "frequency_band": 8,
    "mac_capability_flags": 128,
    "manufacturer_code": 4447,
    "maximum_buffer_size": 127,
    "maximum_incoming_transfer_size": 100,
    "server_mask": 11264,
    "maximum_outgoing_transfer_size": 100,
    "descriptor_capability_field": 0
  },
  "endpoints": {
    "1": {
      "profile_id": "0x0104",
      "device_type": "0x0103",
      "input_clusters": [
        "0x0000",
        "0x0001",
        "0x0003"
      ],
      "output_clusters": [
        "0x0003",
        "0x0006"
      ]
    },
    "2": {
      "profile_id": "0x0104",
      "device_type": "0x0103",
      "input_clusters": [
        "0x0003"
      ],
      "output_clusters": [
        "0x0003",
        "0x0006"
      ]
    },
    "3": {
      "profile_id": "0x0104",
      "device_type": "0x0103",
      "input_clusters": [
        "0x0003"
      ],
      "output_clusters": [
        "0x0006"
      ]
    },
    "4": {
      "profile_id": "",
      "device_type": "",
      "input_clusters": [],
      "output_clusters": []
    },
    "5": {
      "profile_id": "",
      "device_type": "",
      "input_clusters": [],
      "output_clusters": []
    },
    "6": {
      "profile_id": "",
      "device_type": "",
      "input_clusters": [],
      "output_clusters": []
    }
  },
  "manufacturer": "LUMI",
  "model": "lumi.remote.b28ac1"
},
```

## Additional information
<!--
  Please include any additional information that is important to this PR.
  For example, if this PR is a potentially breaking change, mention that here.
  If this PR requires other PRs to be merged in HA Core or other projects, mention that.
  Lastly, if this PR fixes a specific issue, please include "Fixes #xxxx".
-->

I've two switches, one of which seems to be intermittently affected by the issue https://github.com/home-assistant/core/issues/127874 whereby it's recognised and works correctly after initial pairing, but following a restart of the ZHA integration or Home Assistant it's recognised as unk_manufacturer, unk_model and no longer functions.

Update:
I haven't encountered the issue above in the last week with the new v2 quirk (following multiple reboots and zha reloads).

## Checklist
<!--
  Put an 'x' in all boxes that apply.
  Note: You do not need to tick all boxes before creating a PR.
-->

- [x] The changes are tested and work correctly
- [x] `pre-commit` checks pass / the code has been formatted using Black
- [ ] Tests have been added to verify that the new code works

jeverley committed:
Add variant 5 for double rocker switch

jeverley commented:
I'm re-implementing this as a v2 quirk to expose the configuration options as well as exploring whether this resolves the unknown device bug https://github.com/home-assistant/core/issues/127874 - we shall see.

jeverley committed:
Refactor to use v2 quirk format

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

jeverley committed:
Add all OnOff and Identify clusters explicitly

This is to cover all device variants

TheJulianJES commented on zhaquirks/xiaomi/aqara/remote_h1.py:
```diff
@@ -60,336 +42,157 @@
+        manufacturer="Aqara", model="Wireless Remote Switch H1 (Double Rocker)"
+    )
+    .adds(AqaraRemoteManuSpecificCluster)
+    .adds(Identify)
+    .adds(Identify, cluster_type=ClusterType.Client)
+    .adds(Identify, endpoint_id=2)
+    .adds(Identify, endpoint_id=2, cluster_type=ClusterType.Client)
+    .adds(Identify, endpoint_id=3)
+    .adds(Identify, endpoint_id=3, cluster_type=ClusterType.Client)
```
Are we adding all these Identify clusters because the signature reported by the device is sometimes incomplete?
Is there any real function for these on endpoint 2 and 3 (or the client clusters)? If not, I wonder if we can just not add them?

jeverley replied:
I've simplified this in https://github.com/zigpy/zha-device-handlers/pull/3564/commits/76158d1927a1bcad8fbe451a2dbeebbd423f1544 to just add the first on endpoint 1 (it seems to be the only one that actually responds).

TheJulianJES commented on zhaquirks/xiaomi/aqara/remote_h1.py:
```diff
@@ -60,336 +42,157 @@
+    .friendly_name(
+        manufacturer="Aqara", model="Wireless Remote Switch H1 (Double Rocker)"
+    )
+    .adds(AqaraRemoteManuSpecificCluster)
```
IIRC, if a cluster is ever reported by the device, we should use `.replaces`. It also removes the existing cluster with the same cluster id, should there be one already. It should *not* error out if there is no cluster with that ID.
Then, `replaces` also adds the custom cluster.

I don't see it in the `signature` from the v1 quirks, but I think we should still consider using `replaces` here, as it's a real cluster present on the device, just not exposed in the signature (by the device).
I guess that could theoretically change. With `adds`, we only add the custom cluster if one doesn't exist with the same cluster ID already (again, IIRC). In that case, we would not add the attribute definitions.

I think we should just use `replaces` for this one.

jeverley replied:
Agreed, using replace should shield us from future variants which do expose the cluster.
Amended in https://github.com/zigpy/zha-device-handlers/pull/3564/commits/76158d1927a1bcad8fbe451a2dbeebbd423f1544

TheJulianJES commented on zhaquirks/xiaomi/aqara/remote_h1.py:
```diff
@@ -60,336 +42,157 @@
+    .adds(Identify, endpoint_id=2, cluster_type=ClusterType.Client)
+    .adds(Identify, endpoint_id=3)
+    .adds(Identify, endpoint_id=3, cluster_type=ClusterType.Client)
+    .adds(MultistateInputCluster)
+    .adds(MultistateInputCluster, endpoint_id=2)
+    .adds(MultistateInputCluster, endpoint_id=3)
```
Hmm, I guess the same might apply here. These are real clusters, just not exposed in the signature by the current device firmware.
I'd also consider using `replaces` for these.

jeverley replied:
Some variants to expose the cluster in their signature, definitely correct to use replace.
Amended in https://github.com/zigpy/zha-device-handlers/pull/3564/commits/76158d1927a1bcad8fbe451a2dbeebbd423f1544

TheJulianJES commented on zhaquirks/xiaomi/aqara/remote_h1.py:
```diff
@@ -60,336 +42,157 @@
+    .adds(MultistateInputCluster)
+    .adds(MultistateInputCluster, endpoint_id=2)
+    .adds(MultistateInputCluster, endpoint_id=3)
+    .adds(OnOff, cluster_type=ClusterType.Client)
+    .adds(OnOff, endpoint_id=2, cluster_type=ClusterType.Client)
+    .adds(OnOff, endpoint_id=3, cluster_type=ClusterType.Client)
```
These are probably fine to leave as-is. If they're not exposed by the device signature, we add them. If they are, we don't add them, but it's still the same cluster, as it's the default ZCL `OnOff` one.
I believe these clusters are also used if bound to another device when the remote is in that mode.

TheJulianJES commented on zhaquirks/xiaomi/aqara/remote_h1.py:
```diff
@@ -60,336 +42,157 @@
+(
+    QuirkBuilder(LUMI, "lumi.remote.b18ac1")
+    .friendly_name(
+        manufacturer="Aqara", model="Wireless Remote Switch H1 (Single Rocker)"
```
Due to the current implementation we have for `friendly_name`, this might break existing blueprints.
I wonder if we should comment it out for now (until the ZHA implementation is changed?).

jeverley replied:
Have commented these out for now in https://github.com/zigpy/zha-device-handlers/pull/3564/commits/76158d1927a1bcad8fbe451a2dbeebbd423f1544

TheJulianJES reviewed (commented):
Nice work!

jeverley committed:
Incorporate feedback into quirk definition

Addresses feedback in PR:
 - https://github.com/zigpy/zha-device-handlers/pull/3564#discussion_r1865031784
 - https://github.com/zigpy/zha-device-handlers/pull/3564#discussion_r1865032754
 - https://github.com/zigpy/zha-device-handlers/pull/3564#discussion_r1865034058
 - https://github.com/zigpy/zha-device-handlers/pull/3564#discussion_r1865034594
 - https://github.com/zigpy/zha-device-handlers/pull/3564#discussion_r1865034838

jeverley committed:
Introduce zcl_type so that enums can be used to describe attribute values

TheJulianJES commented on zhaquirks/xiaomi/aqara/remote_h1.py:
```diff
@@ -60,336 +42,154 @@
-                OUTPUT_CLUSTERS: [OnOff.cluster_id],
+(
+    QuirkBuilder(LUMI, "lumi.remote.b28ac1")
+    #    .friendly_name(
+    #        manufacturer="Aqara", model="Wireless Remote Switch H1 (Double Rocker)"
+    #    )
```
Let's just add a small comment here about why this is commented out. I hope to double check if we can actually add this soon.
```suggestion
    # temporarily commented out due to potentially breaking existing blueprints
    # .friendly_name(
    #     manufacturer="Aqara", model="Wireless Remote Switch H1 (Double Rocker)"
    # )
```

TheJulianJES commented on zhaquirks/xiaomi/aqara/remote_h1.py:
```diff
@@ -60,336 +42,154 @@
-                OUTPUT_CLUSTERS: [OnOff.cluster_id],
+(
+    QuirkBuilder(LUMI, "lumi.remote.b18ac1")
+    #    .friendly_name(
+    #        manufacturer="Aqara", model="Wireless Remote Switch H1 (Single Rocker)"
+    #    )
```
Let's just add a small comment here about why this is commented out. I hope to double check if we can actually add this soon.
```suggestion
    # temporarily commented out due to potentially breaking existing blueprints
    # .friendly_name(
    #     manufacturer="Aqara", model="Wireless Remote Switch H1 (Single Rocker)"
    # )
```

TheJulianJES reviewed (approved):
Awesome! Let's try this.

TheJulianJES committed:
Add comment on why `friendly_name` is commented out

It's currently commented out due to potentially breaking existing blueprints, as ZHA replaces the model and manufacturer exposed to HA.
